chalice/deploy/newdeployer.py

```diff
@@ -0,0 +1,472 @@
+from chalice import app  # noqa
+from chalice.deploy.packager import LambdaDeploymentPackager
+from chalice.deploy.packager import PipRunner, SubprocessPip
+from chalice.deploy.packager import DependencyBuilder as PipDependencyBuilder
+from chalice.policy import AppPolicyGenerator
+from chalice.constants import LAMBDA_TRUST_POLICY
+from chalice.constants import DEFAULT_LAMBDA_TIMEOUT
+from chalice.constants import DEFAULT_LAMBDA_MEMORY_SIZE
+from chalice.awsclient import TypedAWSClient, ResourceDoesNotExistError
+
+
+def create_default_deployer(session):
+    # type: (Session) -> Deployer
+    client = TypedAWSClient(session)
+    osutils = OSUtils()
+    pip_runner = PipRunner(pip=SubprocessPip(osutils=osutils),
+                           osutils=osutils)
+    dependency_builder = PipDependencyBuilder(
+        osutils=osutils,
+        pip_runner=pip_runner
+    )
+    return Deployer(
+        resource_builder=ApplicationGraphBuilder(),
+        deps_builder=DependencyBuilder(),
+        build_stage=BuildStage(
+            steps=[
+                InjectDefaults(),
+                DeploymentPackager(
+                    packager=LambdaDeploymentPackager(
+                        osutils=osutils,
+                        dependency_builder=dependency_builder,
+                        ui=UI(),
+                    ),
+                ),
+                PolicyGenerator(
+                    policy_gen=AppPolicyGenerator(
+                        osutils=osutils
+                    ),
+                ),
+            ],
+        ),
+        plan_stage=PlanStage(
+            client=client, osutils=osutils,
+        ),
+        executor=Executor(client),
+    )
+
+
+class UnresolvedValueError(Exception):
+    MSG = (
+        "The API parameter '%s' has an unresolved value "
+        "of %s in the method call: %s"
+    )
+
+    def __init__(self, key, value, method_name):
+        # type: (str, models.Placeholder, str) -> None
+        msg = self.MSG % (key, value, method_name)
+        super(UnresolvedValueError, self).__init__(msg)
+        self.key = key
+        self.value = value
+        self.method_name = method_name
+
+
+class Deployer(object):
+    def __init__(self,
+                 resource_builder,  # type: ApplicationGraphBuilder
+                 deps_builder,      # type: DependencyBuilder
+                 build_stage,       # type: BuildStage
+                 plan_stage,        # type: PlanStage
+                 executor,          # type: Executor
+                 ):
+        # type: (...) -> None
+        self._resource_builder = resource_builder
+        self._deps_builder = deps_builder
+        self._build_stage = build_stage
+        self._plan_stage = plan_stage
+        self._executor = executor
+
+    def deploy(self, config, chalice_stage_name):
+        # type: (Config, str) -> Dict[str, Any]
+        application = self._resource_builder.build(config, chalice_stage_name)
```

Comment:
application_graph? I still think of a chalice app object when I see application.

Reply:
Eh, I mean the type of the object *is* a `models.Application`, and the methods do make it clear the type it returns (and accepts).  Also the chalice app is a type `Chalice` not `application`.  I don't see this being confusing given the required type hints and the fact that this is all in the deployer module.  Technically all of these model objects are graphs because they all have a `dependencies()` method.